linkerd/app/outbound/src/switch_logical.rs

```diff
@@ -37,43 +32,46 @@ impl<S> Outbound<S> {
-                        return Ok(svc::Either::A(Endpoint::from_metadata(
-                            addr,
-                            metadata,
-                            no_tls_reason,
+        let stack = endpoint
+            .push_switch(
+                move |(profile, target): (Option<profiles::Receiver>, T)| -> Result<_, Never> {
+                    if let Some(rx) = profile {
+                        let profiles::Profile {
+                            ref addr,
+                            ref endpoint,
                             opaque_protocol,
-                        )));
+                            ..
+                        } = *rx.borrow();
+
+                        // If the profile provides an endpoint, then the target is single endpoint and
+                        // not a logical/load-balanced service.
+                        if let Some((addr, metadata)) = endpoint.clone() {
+                            return Ok(svc::Either::A(Endpoint::from_metadata(
+                                addr,
+                                metadata,
+                                no_tls_reason,
+                                opaque_protocol,
+                            )));
+                        }
+
+                        // Otherwise, if the profile provides a (named) logical address, then we build a
+                        // logical stack so we apply routes, traffic splits, and load balancing.
+                        if let Some(logical_addr) = addr.clone() {
+                            return Ok(svc::Either::B(Logical::new(logical_addr, rx.clone())));
+                        }
                     }
 
-                    // Otherwise, if the profile provides a (named) logical address, then we build a
-                    // logical stack so we apply routes, traffic splits, and load balancing.
-                    if let Some(logical_addr) = addr.clone() {
-                        return Ok(svc::Either::B(Logical::new(logical_addr, rx.clone())));
-                    }
-                }
-
-                // If there was no profile or it didn't include any useful metadata, create a bare
-                // endpoint from the original destination address.
-                Ok(svc::Either::A(Endpoint::forward(
-                    target.param(),
-                    no_tls_reason,
-                )))
-            },
-            logical,
-        );
+                    // If there was no profile or it didn't include any useful metadata, create a bare
+                    // endpoint from the original destination address.
+                    Ok(svc::Either::A(Endpoint::forward(
+                        target.param(),
+                        no_tls_reason,
+                    )))
+                },
+                logical,
+            )
+            .push_on_response(svc::BoxService::layer())
+            .push(svc::BoxNewService::layer());
```

Comment:
also, i was wondering if it would make more sense to put the boxing _before_ the switch here, because boxing after the switch would mean that the `Either` future type still contains both branches of the switch, even if it's immediately erased. but, since this is a method on `Outbound`, and _all_ the other such methods now box, i guess we can guarantee that both stacks are already boxed before calling this...